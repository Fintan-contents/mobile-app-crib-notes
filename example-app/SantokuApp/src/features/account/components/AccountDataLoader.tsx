import {handleErrorWithAlert} from 'bases/core/errors/handleErrorWithAlert';
import {resolveErrorMessage} from 'bases/message/resolveErrorMessage';
import React, {useEffect, useRef} from 'react';

import {useIsLoggedIn} from '../client-states/useIsLoggedIn';
import {useAccountData} from '../services/account/useAccountData';

export const AccountDataLoader: React.FC<React.PropsWithChildren> = ({children}) => {
  const [isLoggedIn] = useIsLoggedIn();
  const isAutoLoggedIn = useRef(isLoggedIn).current;
  const {isLoading, isLoadingError, errorUpdateCount, error} = useAccountData({
    enabled: isLoggedIn,
    meta: {disableGlobalErrorHandler: true},
  });

  useEffect(() => {
    // オートログイン済で初回のデータ取得時にエラーが発生した場合は
    if (isAutoLoggedIn && isLoadingError) {
      if (errorUpdateCount === 1) {
        const {title, message} = resolveErrorMessage(error);
        handleErrorWithAlert(error, title, message);
      }
    }
  }, [error, errorUpdateCount, isAutoLoggedIn, isLoadingError]);

  if (isAutoLoggedIn && (isLoading || isLoadingError)) {
    return null;
  }
  return <>{children}</>;
};
