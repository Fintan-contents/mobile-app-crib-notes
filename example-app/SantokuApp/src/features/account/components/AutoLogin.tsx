import {handleErrorWithAlert} from 'bases/core/errors/handleErrorWithAlert';
import {resolveErrorMessage} from 'bases/message/resolveErrorMessage';
import React, {useEffect} from 'react';

import {useIsLoggedIn} from '../client-states/useIsLoggedIn';
import {isUnauthorizedError} from '../errors/UnauthorizedError';
import {useAuthCommands} from '../services/auth/useAuthCommands';

export const AutoLogin: React.FC<React.PropsWithChildren> = ({children}) => {
  const [isLoggedIn] = useIsLoggedIn();
  const {autoLogin} = useAuthCommands();
  useEffect(() => {
    autoLogin().catch(e => {
      if (!isUnauthorizedError(e)) {
        const {title, message} = resolveErrorMessage(e);
        handleErrorWithAlert(e, title, message);
      }
    });
  }, [autoLogin]);
  // オートログイン時にエラーが発生した場合は、isLoggedInには何も値が設定されない
  // isLoggedInに値が設定されたタイミングでオートログインの処理完了とみなす
  if (isLoggedIn == null) {
    return null;
  }
  return <>{children}</>;
};
