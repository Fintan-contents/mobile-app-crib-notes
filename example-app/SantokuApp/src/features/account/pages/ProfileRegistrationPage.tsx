import {handleErrorWithAlert} from 'bases/core/errors/handleErrorWithAlert';
import {m} from 'bases/message/Message';
import {FilledButton} from 'bases/ui/button/FilledButton';
import {TextInput} from 'bases/ui/input/TextInput';
import {Spacer} from 'bases/ui/spacer/Spacer';
import {TermsOfServiceAgreementStatus} from 'features//backend/apis/model';
import React, {useCallback} from 'react';
import {StyleSheet, View} from 'react-native';

import {isUnauthorizedError} from '../errors/UnauthorizedError';
import {ProfileFormValues, useProfileRegistrationForm} from '../forms/useProfileRegistrationForm';
import {useAccountCommands} from '../services/account/useAccountCommands';
import {clientLogout} from '../services/auth/clientLogout';
import {useAuthCommands} from '../services/auth/useAuthCommands';

export type ProfileRegistrationPageProps = {
  termsAgreementStatus: TermsOfServiceAgreementStatus;
};
export const ProfileRegistrationPage: React.FC<ProfileRegistrationPageProps> = ({termsAgreementStatus}) => {
  const {signup, isSigningUp} = useAuthCommands();
  const {agreeTerms, isAgreeingTerms} = useAccountCommands();
  const onSubmit = useCallback(
    async (values: ProfileFormValues) => {
      try {
        await signup({nickname: values.nickname});
        await agreeTerms(termsAgreementStatus);
      } catch (e) {
        // ここではサインアップに成功したaccountId、passwordを使用してログインしているため、UnauthorizedErrorが発生しない想定です。
        // もし発生した場合は、クライアント側のログアウト処理を実施後、Firebase Crashlyticsにエラーログを送信します。
        if (isUnauthorizedError(e)) {
          await clientLogout();
          handleErrorWithAlert(e, m('app.account.サインアップエラータイトル'), m('app.account.サインアップエラー本文'));
        }
      }
    },
    [agreeTerms, signup, termsAgreementStatus],
  );
  const {form, clearNickname} = useProfileRegistrationForm({
    onSubmit,
  });

  return (
    <View style={styles.container} testID="ProfileRegistration">
      <TextInput
        label={m('ニックネーム')}
        value={form.values.nickname}
        onChangeText={form.handleChange('nickname')}
        showClearButton={!!form.values.nickname}
        onClearButtonPress={clearNickname}
        errorMessage={form.errors.nickname}
      />
      <Spacer heightRatio={0.05} />
      <FilledButton title={m('登録')} onPress={form.submitForm} loading={isSigningUp || isAgreeingTerms} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 30,
  },
});
