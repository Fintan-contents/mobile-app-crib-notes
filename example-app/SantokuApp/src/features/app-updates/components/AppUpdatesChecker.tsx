import * as Application from 'expo-application';
import React, {useEffect} from 'react';

import {useAppUpdates} from '../services/useAppUpdates';
import {showUpdateRequiredDialog} from '../use-cases/showUpdateRequiredDialog';

const AppUpdatesChecker: React.FC<React.PropsWithChildren> = ({children}) => {
  const {appUpdates} = useAppUpdates();

  useEffect(() => {
    if (appUpdates?.updateRequired) {
      showUpdateRequiredDialog(Application.nativeApplicationVersion!);
    }
  }, [appUpdates]);
  if (appUpdates === undefined || appUpdates.updateRequired) {
    return null;
  }

  return <>{children}</>;
};

export {AppUpdatesChecker};
