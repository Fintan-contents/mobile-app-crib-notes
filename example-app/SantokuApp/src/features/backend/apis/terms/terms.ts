/**
 * Generated by orval v6.10.2 üç∫
 * Do not edit manually.
 * SantokuApp REST API
 * SantokuApp„ÅÆREST API„ÄÇ
 * OpenAPI spec version: 1.0.0
 */
import {useQuery} from 'react-query';
import type {UseQueryOptions, QueryFunction, UseQueryResult, QueryKey} from 'react-query';
import type {TermsOfService} from '.././model';
import {backendCustomInstance} from '../../utils/customInstance';
import type {ErrorType} from '../../utils/customInstance';

type AwaitedInput<T> = PromiseLike<T> | T;

type Awaited<O> = O extends AwaitedInput<infer T> ? T : never;

/**
 * ÊúâÂäπ„Å™Âà©Áî®Ë¶èÁ¥Ñ„ÇíÂèñÂæó„Åó„Åæ„Åô„ÄÇ„Åì„ÅÆAPI„ÅÆÂëº„Å≥Âá∫„Åó„Å´„ÅØË™çË®ºÊÉÖÂ†±„ÅØ‰∏çË¶Å„Åß„Åô„ÄÇ

 * @summary ÊúâÂäπ„Å™Âà©Áî®Ë¶èÁ¥Ñ„ÅÆÂèñÂæó
 */
export const getTerms = (signal?: AbortSignal) => {
  return backendCustomInstance<TermsOfService>({url: `/terms`, method: 'get', signal});
};

export const getGetTermsQueryKey = () => [`/terms`];

export type GetTermsQueryResult = NonNullable<Awaited<ReturnType<typeof getTerms>>>;
export type GetTermsQueryError = ErrorType<unknown>;

export const useGetTerms = <TData = Awaited<ReturnType<typeof getTerms>>, TError = ErrorType<unknown>>(options?: {
  query?: UseQueryOptions<Awaited<ReturnType<typeof getTerms>>, TError, TData>;
}): UseQueryResult<TData, TError> & {queryKey: QueryKey} => {
  const {query: queryOptions} = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetTermsQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getTerms>>> = ({signal}) => getTerms(signal);

  const query = useQuery<Awaited<ReturnType<typeof getTerms>>, TError, TData>(
    queryKey,
    queryFn,
    queryOptions,
  ) as UseQueryResult<TData, TError> & {queryKey: QueryKey};

  query.queryKey = queryKey;

  return query;
};
