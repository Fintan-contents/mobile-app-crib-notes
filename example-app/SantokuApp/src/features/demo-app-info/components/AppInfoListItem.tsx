import React from 'react';
import {Text, View} from 'react-native';
import {ListItem} from 'react-native-elements';

export type AppInfoItemData = {
  name?: string;
  value: string | null;
};

type AppInfoListItemProps = {
  item: AppInfoItemData;
};

export const AppInfoListItem: React.FC<AppInfoListItemProps> = ({item}) => {
  return !item.name ? null : (
    <ListItem bottomDivider>
      <ListItem.Content>
        <ListItem.Title>
          <Text>{item.name}</Text>
        </ListItem.Title>
        <View style={{marginTop: 10, marginLeft: 10}}>
          <Text>{item.value}</Text>
        </View>
      </ListItem.Content>
    </ListItem>
  );
};
