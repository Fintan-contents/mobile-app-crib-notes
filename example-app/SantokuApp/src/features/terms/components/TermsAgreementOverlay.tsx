import {FullWindowOverlay} from 'bases/ui/overlay/FullWindowOverlay';
import React, {useCallback, useMemo, useState} from 'react';

import {
  TermsAgreementOverlayComponent,
  TermsAgreementOverlayProps,
  TermsAgreementOverlayShowProps,
} from './TermsAgreementOverlayComponent';

type TermsAgreementOverlayType = {
  show: (props: TermsAgreementOverlayShowProps) => void;
  close: () => void;
  Component: typeof Component;
};

/**
 * 利用規約をOverlay表示するコンポーネント。
 */
const Component: React.FC = () => {
  const [state, setState] = useState<
    Omit<TermsAgreementOverlayProps, 'termsOfService'> & Partial<Pick<TermsAgreementOverlayProps, 'termsOfService'>>
  >({
    visible: false,
    close: () => {},
  });
  const close = useCallback(() => setState(prevState => ({...prevState, visible: false})), []);
  const show = useCallback(
    (props: Omit<TermsAgreementOverlayProps, 'visible' | 'close'>) => {
      setState({
        ...props,
        visible: true,
        close,
      });
    },
    [close],
  );

  TermsAgreementOverlay = useMemo(
    () => ({
      ...TermsAgreementOverlay,
      show,
      close,
    }),
    [close, show],
  );

  return (
    <FullWindowOverlay>
      {state.termsOfService && <TermsAgreementOverlayComponent {...state} termsOfService={state.termsOfService} />}
    </FullWindowOverlay>
  );
};

let TermsAgreementOverlay: TermsAgreementOverlayType = {
  show: (props: TermsAgreementOverlayShowProps) => {
    throw new Error('TermsAgreementOverlay.Component is required.');
  },
  close: () => {
    throw new Error('TermsAgreementOverlay.Component is required.');
  },
  Component,
};

export {TermsAgreementOverlay};
