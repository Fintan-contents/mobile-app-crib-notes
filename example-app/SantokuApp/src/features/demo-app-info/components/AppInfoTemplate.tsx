import React from 'react';
import {FlatList} from 'react-native';

import {AppInfoItemData, AppInfoListItem} from './AppInfoListItem';

type AppInfoTemplateProps = {
  items: AppInfoItemData[];
  keyExtractor: (item: AppInfoItemData, index: number) => string;
  testID?: string;
};

export const AppInfoTemplate: React.FC<AppInfoTemplateProps> = ({items, keyExtractor, testID}) => {
  return <FlatList data={items} renderItem={AppInfoListItem} keyExtractor={keyExtractor} testID={testID} />;
};
