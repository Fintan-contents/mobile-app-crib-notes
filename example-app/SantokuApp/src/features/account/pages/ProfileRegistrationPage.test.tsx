import {render, screen} from '@testing-library/react-native';
import {BundledMessagesLoader} from 'bases/message/BundledMessageLoader';
import {loadMessages} from 'bases/message/Message';
import {Snackbar} from 'bases/ui/snackbar/Snackbar';
import {AppThemeProvider} from 'bases/ui/theme/AppThemeProvider';
import {enhanceValidator} from 'bases/validator';
import {useGetTerms} from 'features/backend/apis/terms/terms';
import {TermsAgreementOverlay} from 'features/terms/components/TermsAgreementOverlay';
import React from 'react';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import {QueryClient, QueryClientProvider} from 'react-query';

import {ProfileRegistrationPageProps} from './ProfileRegistrationPage';

jest.mock('features/backend/apis/terms/terms');

const Wrapper: React.FC<React.PropsWithChildren> = ({children}) => {
  const queryClient = new QueryClient();
  return (
    <SafeAreaProvider>
      <AppThemeProvider>
        <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
        <TermsAgreementOverlay.Component />
        <Snackbar.Component />
      </AppThemeProvider>
    </SafeAreaProvider>
  );
};

beforeAll(async () => {
  await loadMessages(new BundledMessagesLoader());
  enhanceValidator();
});

describe('ProfileRegistrationScreen', () => {
  it('マウントされたときに正常にレンダリングされること', () => {
    (useGetTerms as jest.Mock).mockReturnValue({refetch: () => {}});
    // importでLoginScreenを読み込むと、メッセージのロードが完了する前にメッセージを読み込んでしまうため、requireで取得する
    // requireした場合の型はanyとなってしまいESLintエラーが発生しますが無視します。
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const Page = require('./ProfileRegistrationPage').ProfileRegistrationPage as React.FC<ProfileRegistrationPageProps>;
    render(<Page termsAgreementStatus={{hasAgreed: true, agreedVersion: '1.0.0'}} />, {
      wrapper: Wrapper,
    });
    expect(screen.queryByTestId('ProfileRegistration')).not.toBeNull();
    expect(screen).toMatchSnapshot();
  });
});
