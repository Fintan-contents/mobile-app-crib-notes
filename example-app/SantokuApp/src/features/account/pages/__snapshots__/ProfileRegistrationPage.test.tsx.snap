// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileRegistrationScreen マウントされたときに正常にレンダリングされること 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 30,
      }
    }
    testID="ProfileRegistration"
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#212935",
              "fontSize": 15,
              "fontWeight": "normal",
            }
          }
        >
          ニックネーム
        </Text>
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#9e9e9e",
              "borderBottomWidth": 2,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "paddingVertical": 8,
              "transform": [
                {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            editable={true}
            onChangeText={[Function]}
            placeholderTextColor="#86939e"
            style={
              {
                "color": "#212935",
                "flex": 1,
                "fontSize": 17,
                "includeFontPadding": false,
                "lineHeight": 24,
                "minHeight": 0,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "height": 32,
          "width": 0,
        }
      }
      testID="Spacer"
    />
    <View
      style={
        [
          {
            "borderRadius": 3,
            "overflow": "hidden",
          },
          {
            "borderRadius": 3,
          },
          {
            "borderRadius": 8,
            "width": 90,
          },
          false,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#3a6be8",
              "borderColor": "#2089dc",
              "borderRadius": 8,
              "borderWidth": 0,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "center",
              "padding": 8,
              "width": 90,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            登録
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <View
    testID="FullWindowOverlay"
  />,
  <View
    testID="FullWindowOverlay"
  />,
]
`;
